### PR TITLE
forbid setattr type changes and enable block replacement

### PR DESCRIPTION
**Pull request by szha**
this change includes:
1. prevent attributes whose names don't start with '_' from changing types
2. properly replace _children if both existing value and new value are blocks.

### CODE REVIEW

File: python/mxnet/gluon/block.py [hunk 1]
```diff
@@ -172,9 +172,23 @@ def __repr__(self):
...
+        if hasattr(self, name):
+            existing = getattr(self, name)
+            if not name.startswith('_') and not isinstance(value, type(existing)):
+                raise TypeError('Changing attribute type for {name} from {type1} to {type2}' \
+                                'is not allowed.'.format(name=name,
+                                                         type1=type(existing),
+                                                         type2=type(value)))
+            if isinstance(existing, Block):
+                for i, c in enumerate(self._children):
+                    if c == existing:
```
Comment (piiswrong) [inline python/mxnet/gluon/block.py]: is instead of ==

**Comment by piiswrong** [comment]
Please add tests for these new behaviors

File: tests/python/unittest/test_gluon.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
+import unittest
```
Comment (piiswrong) [inline tests/python/unittest/test_gluon.py]: ?

Comment (piiswrong) [reply tests/python/unittest/test_gluon.py]: use nosetest

File: python/mxnet/gluon/block.py [hunk 3]
```diff
@@ -172,9 +172,23 @@ def __repr__(self):
 
     def __setattr__(self, name, value):
         """Registers parameters."""
+
+        if hasattr(self, name):
+            existing = getattr(self, name)
+            if not name.startswith('_') and not isinstance(value, type(existing)):
+                raise TypeError('Changing attribute type for {name} from {type1} to {type2}' \
```
Comment (piiswrong) [inline python/mxnet/gluon/block.py]: Let's restrict the check to Parameter and Block.

Comment (szha) [reply python/mxnet/gluon/block.py]: Existing attribute value can be NoneType and later changes to Block or Parameter. Should this be allowed?

**Pull request merged**